samza-kv/src/test/scala/org/apache/samza/storage/kv/TestKeyValueStorageEngine.scala

```diff
@@ -163,6 +167,50 @@ class TestKeyValueStorageEngine {
     assertEquals(15, metrics.restoredBytesGauge.getValue) // 3 keys * 2 bytes/key +  3 msgs * 3 bytes/msg
   }
 
+  @Test
+  def testRestoreInterruptedThrowsInterruptException(): Unit = {
+    val changelogSSP = new SystemStreamPartition("TestSystem", "TestStream", new Partition(0))
+    val iterator = mock(classOf[ChangelogSSPIterator])
+    val executorService = Executors.newSingleThreadExecutor()
+    val restore = new Callable[Void] {
+      override def call(): Void = {
+        engine.restore(iterator)
+        null
+      }
+    }
+
+    when(iterator.hasNext)
+      .thenReturn(true)
+      .thenReturn(true)
+      .thenAnswer(new Answer[Boolean] {
+        override def answer(invocation: InvocationOnMock): Boolean = {
+          executorService.shutdownNow()
+          true
+        }
+      })
+      .thenReturn(false)
+    when(iterator.next())
+      .thenReturn(new IncomingMessageEnvelope(changelogSSP, "0", Array[Byte](1, 2), Array[Byte](3, 4, 5)))
+      .thenReturn(new IncomingMessageEnvelope(changelogSSP, "1", Array[Byte](2, 3), Array[Byte](4, 5, 6)))
+      .thenReturn(new IncomingMessageEnvelope(changelogSSP, "2", Array[Byte](3, 4), Array[Byte](5, 6, 7)))
+    when(iterator.getMode)
+      .thenReturn(Mode.RESTORE)
+      .thenReturn(Mode.RESTORE)
+      .thenReturn(Mode.RESTORE)
+
+    try {
+      val restoreFuture = executorService.submit(restore)
+      restoreFuture.get()
```

Comment:
Should an `Assert.fail()` go after the `get()`? It looks like if the `restoreFuture.get()` just succeeds normally, then this test would also pass.

Reply:
Good catch 👍